### PR TITLE
Refactor annotations parameter to accept explicit values

### PR DESCRIPTION
**Pull request by karenzshea**
# Issue

Closes https://github.com/Project-OSRM/osrm-backend/issues/3563

## Tasklist
-  [ ] ~~Adapt annotations parsing for map matching route~~ matching inherits from route
 - [x] Abstract annotation insertion to steps into a function
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [x] add cucumber cases (see docs/testing.md)
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

File: features/testbot/matching.feature [hunk 1]
```diff
@@ -141,9 +167,9 @@ Feature: Basic Map Matching
         And the contract extra arguments "--segment-speed-file {speeds_file}"
 
         When I match I should get
-            | trace | matchings | annotation                                                                                                      |
-            | abeh  | abeh      | 1:1:10.008842:1,0:0:0:0,1:1:10.008842:0,1:1:10.008842:0,1:1:10.008842:0,0:0:0:0,2:2:19.906475:0,1:1:10.008842:0 |
-            | abci  | abci      | 1:1:10.008842:1,0:0:0:0,1:1:10.008842:0,0:0:0:0,1:1:10.010367:0                                                 |
+            | trace | matchings | annotation                                                                                                                      |
+            | abeh  | abeh      | 1:10.008842:1:1:1,0:0:0:2:0,1:10.008842:1:2:0,1:10.008842:1:3:0,1:10.008842:1:4:0,0:0:0:4:0,2:19.906475:2:5:0,1:10.008842:1:6:0 |
+            | abci  | abci      | 1:10.008842:1:1:1,0:0:0:2:0,1:10.008842:1:2:0,0:0:0:2:0,1:10.010367:1:3:0                                                       |
```
Comment (daniel-j-h) [inline features/testbot/matching.feature]: What are the changes here?

File: features/testbot/matching.feature [hunk 2]
```diff
@@ -118,6 +118,32 @@ Feature: Basic Map Matching
...
+        And the speed file
+        """
+        1,2,36
+        """
+
+        And the contract extra arguments "--segment-speed-file {speeds_file}"
+
+        When I match I should get
+            | trace | matchings | annotation    |
+            | ach   | ach       | 1,1,0,1,1,2,1 |
```
Comment (daniel-j-h) [inline features/testbot/matching.feature]: The annotations here are durations, right? How do I read those in the context of the node map? Why is there a duration 0 in there?

File: features/testbot/matching.feature [hunk 3]
```diff
@@ -337,3 +363,29 @@ Feature: Basic Map Matching
...
+            a - b
+                |
+                c
+            """
+
+        And the nodes
+            | node | id |
+            | a    | 1  |
+            | b    | 2  |
+            | c    | 3  |
```
Comment (daniel-j-h) [inline features/testbot/matching.feature]: Why is this needed?

File: include/engine/api/route_api.hpp [hunk 4]
```diff
@@ -215,43 +215,71 @@ class RouteAPI : public BaseAPI
...
-                              [this, &nodes](const OSMNodeID &node_id) {
-                                  nodes.values.push_back(static_cast<std::uint64_t>(node_id));
-                              });
                 util::json::Object annotation;
-                annotation.values["distance"] = std::move(distances);
-                annotation.values["duration"] = std::move(durations);
-                annotation.values["weight"] = std::move(weights);
-                annotation.values["nodes"] = std::move(nodes);
-                annotation.values["datasources"] = std::move(datasources);
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Duration)
```
Comment (daniel-j-h) [inline include/engine/api/route_api.hpp]: Hm either you do the bit fiddling here or what about abstracting it away? The code here could work no matter how the `AnnotationsType` is implemented.

File: include/engine/api/route_api.hpp [hunk 5]
```diff
@@ -215,43 +215,71 @@ class RouteAPI : public BaseAPI
...
+                {
+                    util::json::Array nodes;
+                    nodes.values.reserve(leg_geometry.osm_node_ids.size());
+                    std::for_each(leg_geometry.osm_node_ids.begin(),
+                                  leg_geometry.osm_node_ids.end(),
+                                  [this, &nodes](const OSMNodeID &node_id) {
+                                      nodes.values.push_back(static_cast<std::uint64_t>(node_id));
+                                  });
+                    annotation.values["nodes"] = std::move(nodes);
+                }
```
Comment (daniel-j-h) [inline include/engine/api/route_api.hpp]: Hm the ~100 lines above are all doing the same:
- check if annotation x was requested
- create `json::Array` for many x
- extract `step.x` from `RouteStep` and add into array
- add array to annotations response as `"x" : [...]`

I think you could abstract over this and make it extensible for the future at the same time.

What about abstracting this behind a function which takes
- a `AnnotationsType`
- a lambda taking a `RouteStep` and returning annotations items out of it (e.g. `step.duration` for `AnnotationsType::Duration`
- a string describing the annotations type for the json object's key as in `"key": [...]` above

Example callsite:

```c++
makeAnnotation(AnnotationsType::Duration, "duration", [](auto& step){ return step.duration });
```

which then does all of the steps above behind the scenes and wires the parts together.

File: include/engine/api/route_parameters.hpp [hunk 6]
```diff
@@ -67,6 +67,16 @@ struct RouteParameters : public BaseParameters
         Full,
         False
     };
+    enum class AnnotationsType : int
```
Comment (daniel-j-h) [inline include/engine/api/route_parameters.hpp]: Why the explicit `: int` here? Should be `int` by default.

File: include/engine/api/route_parameters.hpp [hunk 7]
```diff
@@ -94,20 +104,54 @@ struct RouteParameters : public BaseParameters
...
+    RouteParameters(const bool steps_,
+                    const bool alternatives_,
+                    const AnnotationsType annotations_,
+                    const GeometriesType geometries_,
+                    const OverviewType overview_,
+                    const boost::optional<bool> continue_straight_,
+                    Args... args_)
+        : BaseParameters{std::forward<Args>(args_)...}, steps{steps_}, alternatives{alternatives_},
+          annotations_type{annotations_}, annotations{true}, geometries{geometries_},
+          overview{overview_}, continue_straight{continue_straight_}
```
Comment (daniel-j-h) [inline include/engine/api/route_parameters.hpp]: Nice :+1: This should keep the API stable for the time being. We may want to ticket the removal of all those hacks for v6. I think we already have tickets for these kinds of hacks.

File: include/server/api/route_parameters_grammar.hpp [hunk 8]
```diff
@@ -50,17 +57,25 @@ struct RouteParametersGrammar : public BaseParametersGrammar<Iterator, Signature
             "full", engine::api::RouteParameters::OverviewType::Full)(
             "false", engine::api::RouteParameters::OverviewType::False);
 
+        annotations_type.add("true", engine::api::RouteParameters::AnnotationsType::All)(
+            "false", engine::api::RouteParameters::AnnotationsType::None)(
+            "duration", engine::api::RouteParameters::AnnotationsType::Duration)(
+            "nodes", engine::api::RouteParameters::AnnotationsType::Nodes)(
+            "distance", engine::api::RouteParameters::AnnotationsType::Distance)(
+            "weight", engine::api::RouteParameters::AnnotationsType::Weight)(
+            "datasources", engine::api::RouteParameters::AnnotationsType::Datasources);
```
Comment (daniel-j-h) [inline include/server/api/route_parameters_grammar.hpp]: Hm we need a mapping `AnnotationsType` -> `string` here, too. The same is needed for the response generation (as commented above). Maybe we should add this mapping already next to the `AnnotationsType` enum class definition. This would guarantee that we're in sync with name and enum item.

File: include/server/api/route_parameters_grammar.hpp [hunk 9]
```diff
@@ -50,17 +57,25 @@ struct RouteParametersGrammar : public BaseParametersGrammar<Iterator, Signature
...
-            (qi::lit("annotations=") >
-             qi::bool_[ph::bind(&engine::api::RouteParameters::annotations, qi::_r1) = qi::_1]) |
             (qi::lit("geometries=") >
              geometries_type[ph::bind(&engine::api::RouteParameters::geometries, qi::_r1) =
                                  qi::_1]) |
             (qi::lit("overview=") >
-             overview_type[ph::bind(&engine::api::RouteParameters::overview, qi::_r1) = qi::_1]);
+             overview_type[ph::bind(&engine::api::RouteParameters::overview, qi::_r1) = qi::_1]) |
+            (qi::lit("annotations=") >
+             annotations_type[ph::bind(add_annotation, qi::_r1, qi::_1)] % ',');
```
Comment (daniel-j-h) [inline include/server/api/route_parameters_grammar.hpp]: Is an empty list allowed by this? (I don't know => make a unit test)

File: unit_tests/server/parameters_parser.cpp [hunk 10]
```diff
@@ -321,6 +325,37 @@ BOOST_AUTO_TEST_CASE(valid_route_urls)
...
+        parseParameters<RouteParameters>("1,2;3,4?geometries=polyline&"
+                                         "overview=simplified&annotations=duration,weight,nodes");
+    BOOST_CHECK(result_15);
+    BOOST_CHECK_EQUAL(reference_15.geometries, result_15->geometries);
+    BOOST_CHECK_EQUAL(
+        static_cast<bool>(result_2->annotations_type & (RouteParameters::AnnotationsType::Duration |
+                                                        RouteParameters::AnnotationsType::Weight |
+                                                        RouteParameters::AnnotationsType::Nodes)),
+        true);
+    BOOST_CHECK_EQUAL(result_15->annotations, true);
```
Comment (daniel-j-h) [inline unit_tests/server/parameters_parser.cpp]: Order should not matter - care to create a test expressing this?
Also edge cases such as
- empty list
- all possibilities
- all possibilities including All, None and combinations

File: include/engine/api/route_api.hpp (hunk 5 above)
Comment (karenzshea) [inline include/engine/api/route_api.hpp]: :+1: refactoring this part is already in the OP as the todo item `Abstract annotation insertion to steps into a function`

I was going to take a template based approach to the function, but I can consider your suggestion as possible solution as well.

File: include/server/api/route_parameters_grammar.hpp (hunk 8 above)
Comment (oxidase) [inline include/server/api/route_parameters_grammar.hpp]: using karma or a map value->string? would it make sense to make json output completely in karma?

File: include/engine/api/route_parameters.hpp [hunk 11]
```diff
@@ -67,6 +67,16 @@ struct RouteParameters : public BaseParameters
         Full,
         False
     };
+    enum class AnnotationsType : int
+    {
+        None = 0,
+        Duration = 1 << 1,
```
Comment (oxidase) [inline include/engine/api/route_parameters.hpp]: here it can start from `1<<0`, but better to use explicit  numerals 0x01, 0x02, ...

File: include/server/api/route_parameters_grammar.hpp (hunk 8 above)
Comment (daniel-j-h) [inline include/server/api/route_parameters_grammar.hpp]: a value->string mapping would be good enough for now - our json handling is all based on a variant, not sure how karma fits in there

File: include/server/api/route_parameters_grammar.hpp (hunk 9 above)
Comment (karenzshea) [inline include/server/api/route_parameters_grammar.hpp]: Added! to the parameters_parser unit test

File: features/support/shared_steps.js [hunk 12]
```diff
@@ -133,9 +133,19 @@ module.exports = function () {
...
+                                let whitelist = ['duration', 'distance', 'datasources', 'nodes', 'weight'];
+                                if (k.match(/^a:/)) {
+                                    let a_type = k.slice(2);
+                                    if (whitelist.indexOf(a_type) == -1)
+                                        return cb(new Error('Unrecognized annotation field', a_type));
+                                    if (!annotation[a_type])
+                                        return cb(new Error('Annotation not found in response', a_type));
+                                    got[k] = annotation[a_type];
+                                }
+                            });
```
Comment (daniel-j-h) [inline features/support/shared_steps.js]: Hm this is dup. code with above

File: include/engine/api/route_api.hpp [hunk 13]
```diff
@@ -79,6 +79,19 @@ class RouteAPI : public BaseAPI
...
     }
 
+    template <typename GetFn>
+    util::json::Array GetAnnotations(const guidance::LegGeometry &leg, GetFn Get) const
+    {
+        util::json::Array annotations_store;
+        annotations_store.values.reserve(leg.annotations.size());
+        std::for_each(leg.annotations.begin(),
+                      leg.annotations.end(),
+                      [Get, &annotations_store](const guidance::LegGeometry::Annotation &step) {
```
Comment (daniel-j-h) [inline include/engine/api/route_api.hpp]: I think you can simply say `const auto& step` here

File: include/engine/api/route_api.hpp [hunk 14]
```diff
@@ -215,43 +228,54 @@ class RouteAPI : public BaseAPI
...
-                annotation.values["weight"] = std::move(weights);
-                annotation.values["nodes"] = std::move(nodes);
-                annotation.values["datasources"] = std::move(datasources);
+
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Duration)
+                {
+                    annotation.values["duration"] =
+                        GetAnnotations(leg_geometry,
+                                       std::bind(&guidance::LegGeometry::Annotation::duration,
+                                                 std::placeholders::_1));
```
Comment (daniel-j-h) [inline include/engine/api/route_api.hpp]: Instead of bind you can simply use a lambda, which is probably easier to read and does not have all the weird quirks bind has (see https://www.youtube.com/watch?v=zt7ThwVfap0). E.g.

```cpp
[](const auto& v) { return v.distance };
```

File: features/step_definitions/matching.js [hunk 15]
```diff
@@ -111,9 +109,19 @@ module.exports = function () {
                         got.duration = duration.toString();
                     }
 
-                    if (headers.has('annotation')) {
-                        got.annotation = annotation.toString();
-                    }
+                    // if header matches 'a:*', parse out the values for *
+                    // and return in that header
+                    headers.forEach((k, v) => {
```
Comment (oxidase) [inline features/step_definitions/matching.js]: :+1: but lint complains about unused `v` variable

File: features/support/shared_steps.js [hunk 16]
```diff
@@ -133,9 +133,19 @@ module.exports = function () {
                                 got.locations = (locations || '').trim();
                             }
 
-                            if (headers.has('annotation')){
-                                got.annotation = (annotation || '').trim();
-                            }
+                            // if header matches 'a:*', parse out the values for *
+                            // and return in that header
+                            headers.forEach((k, v) => {
```
Comment (oxidase) [inline features/support/shared_steps.js]: and here `v` is unused

File: unit_tests/server/parameters_parser.cpp [hunk 17]
```diff
@@ -128,6 +132,9 @@ BOOST_AUTO_TEST_CASE(valid_route_urls)
     CHECK_EQUAL_RANGE(reference_2.radiuses, result_2->radiuses);
     CHECK_EQUAL_RANGE(reference_2.coordinates, result_2->coordinates);
     CHECK_EQUAL_RANGE(reference_2.hints, result_2->hints);
+    BOOST_CHECK_EQUAL(
+        static_cast<bool>(result_2->annotations_type & RouteParameters::AnnotationsType::All),
+        true);
```
Comment (daniel-j-h) [inline unit_tests/server/parameters_parser.cpp]: Hm maybe add `operator==` overload then? We already have `|` and `&`.

File: unit_tests/server/parameters_parser.cpp [hunk 18]
```diff
@@ -321,6 +328,58 @@ BOOST_AUTO_TEST_CASE(valid_route_urls)
...
+    RouteParameters reference_17{};
+    reference_17.annotations_type = RouteParameters::AnnotationsType::All;
+    reference_17.coordinates = coords_1;
+    auto result_17 = parseParameters<RouteParameters>(
+        "1,2;3,4?overview=simplified&annotations=duration,weight,nodes,datasources,distance");
+    BOOST_CHECK(result_17);
+    BOOST_CHECK_EQUAL(reference_17.geometries, result_17->geometries);
+    BOOST_CHECK_EQUAL(
+        static_cast<bool>(result_2->annotations_type & RouteParameters::AnnotationsType::All),
+        true);
```
Comment (daniel-j-h) [inline unit_tests/server/parameters_parser.cpp]: Same for all of these tests, `type == All`

File: include/engine/api/route_parameters.hpp [hunk 19]
```diff
@@ -94,20 +104,54 @@ struct RouteParameters : public BaseParameters
...
+                    Args... args_)
+        : BaseParameters{std::forward<Args>(args_)...}, steps{steps_}, alternatives{alternatives_},
+          annotations_type{annotations_}, annotations{true}, geometries{geometries_},
+          overview{overview_}, continue_straight{continue_straight_}
     {
     }
 
     bool steps = false;
     bool alternatives = false;
     bool annotations = false;
```
Comment (oxidase) [inline include/engine/api/route_parameters.hpp]: this is now a duplication, better
```
    bool annotations() const { return annotations_type != AnnotationsType::None; };
```

But it is not used in code base anymore, the best would be delete it and update unit tests.

File: include/server/api/route_parameters_grammar.hpp [hunk 20]
```diff
@@ -42,6 +42,13 @@ struct RouteParametersGrammar : public BaseParametersGrammar<Iterator, Signature
 
     RouteParametersGrammar(qi::rule<Iterator, Signature> &root_rule_) : BaseGrammar(root_rule_)
     {
+        const auto add_annotation = [](engine::api::RouteParameters &route_parameters,
```
Comment (oxidase) [inline include/server/api/route_parameters_grammar.hpp]: this lambda is not needed with `operator |=` and in-place phoenix bind (i added related comments)

File: include/engine/api/route_parameters.hpp [hunk 21]
```diff
@@ -94,20 +104,54 @@ struct RouteParameters : public BaseParameters
...
+    template <typename... Args>
+    RouteParameters(const bool steps_,
+                    const bool alternatives_,
+                    const AnnotationsType annotations_,
+                    const GeometriesType geometries_,
+                    const OverviewType overview_,
+                    const boost::optional<bool> continue_straight_,
+                    Args... args_)
+        : BaseParameters{std::forward<Args>(args_)...}, steps{steps_}, alternatives{alternatives_},
+          annotations_type{annotations_}, annotations{true}, geometries{geometries_},
```
Comment (oxidase) [inline include/engine/api/route_parameters.hpp]: here `annotations` is `true` even if `annotations_` is `None`

File: include/server/api/route_parameters_grammar.hpp (hunk 9 above)
Comment (oxidase) [inline include/server/api/route_parameters_grammar.hpp]: ... here in-place functor
```
annotations_type[ph::bind(&engine::api::RouteParameters::annotations_type, qi::_r1) |= qi::_1] % ',');
```

File: include/engine/api/route_parameters.hpp [hunk 22]
```diff
@@ -94,20 +104,54 @@ struct RouteParameters : public BaseParameters
...
+}
+
+inline RouteParameters::AnnotationsType operator|(RouteParameters::AnnotationsType lhs,
+                                                  RouteParameters::AnnotationsType rhs)
+{
+    return (RouteParameters::AnnotationsType)(
+        static_cast<std::underlying_type_t<RouteParameters::AnnotationsType>>(lhs) |
+        static_cast<std::underlying_type_t<RouteParameters::AnnotationsType>>(rhs));
+}
+
```
Comment (oxidase) [inline include/engine/api/route_parameters.hpp]: ... and here compound operator
```
inline RouteParameters::AnnotationsType& operator|=(RouteParameters::AnnotationsType& lhs,
                                                    RouteParameters::AnnotationsType rhs)
{
    return lhs = lhs | rhs;
}
```

File: include/engine/api/route_parameters.hpp (hunk 19 above)
Comment (daniel-j-h) [inline include/engine/api/route_parameters.hpp]: We can't do that - the annotations member attribute is in our public API - we can't change it to a function :(

File: include/server/api/route_parameters_grammar.hpp (hunk 20 above)
Comment (oxidase) [inline include/server/api/route_parameters_grammar.hpp]: due to https://github.com/Project-OSRM/osrm-backend/pull/3628#discussion_r99120126 both members must be updated, so `add_annotation` must remain

File: include/server/api/route_parameters_grammar.hpp [hunk 23]
```diff
@@ -42,6 +42,13 @@ struct RouteParametersGrammar : public BaseParametersGrammar<Iterator, Signature
...
+                                       engine::api::RouteParameters::AnnotationsType &route_param) {
+            route_parameters.annotations_type = route_parameters.annotations_type | route_param;
+            route_parameters.annotations = route_parameters.annotations_type != engine::api::RouteParameters::AnnotationsType::None;
+            return route_parameters;
```
Comment (oxidase) [inline include/server/api/route_parameters_grammar.hpp]: returned value is not used and `return` can be removed

File: include/engine/api/route_parameters.hpp (hunk 19 above)
Comment (karenzshea) [inline include/engine/api/route_parameters.hpp]: https://github.com/Project-OSRM/osrm-backend/issues/3644

**Comment by TheMarex** [comment]
This is still missing the `speed` parameter. Capturing this in https://github.com/Project-OSRM/osrm-backend/issues/3664 to be fixed in another PR.

Otherwise ready to merge.
